execution/src/worker.rs

```diff
@@ -93,34 +126,76 @@ impl ExecutionWorker {
                 // Process management commands
                 cmd = self.controller_manager_rx.recv() => {
                     match cmd {
-                    None => break,
-                    Some(_) => {}
-                }}
+                        None => break,
+                        Some(_) => {}
+                    }
+                },
 
                 // Process commands
                 Some(cmd) = self.controller_command_rx.recv() => self.process_command(cmd).await?,
             }
         }
-        // end loop
+
+        // Shutdown VM, cancel all pending execution requests
+        {
+            let (queue_lock, condvar) = &*self.execution_queue;
+            let mut queue_guard = queue_lock.lock().unwrap();
+            (*queue_guard).push_front(ExecutionRequest::Stop);
+            condvar.notify_all();
+        }
+        let _ = self.vm_join_handle.join();
+
         Ok(())
     }
 
-    fn apply_final_slot(
-        &mut self,
-        slot: Slot,
-        opt_block: Option<(BlockId, &Block)>,
-    ) -> Result<(), ExecutionError> {
-        // TODO
-        Ok(())
+    // asks the VM to reset to its final
+    fn vm_reset(&mut self) {
+        let (queue_lock, condvar) = &*self.execution_queue;
+        let mut queue_guard = queue_lock.lock().unwrap();
+        // cancel all non-final, non-stop requests
+        // Final execution requests are left to maintain final state consistency
+        (*queue_guard).retain(|req| match req {
+            ExecutionRequest::RunFinalStep(..) => true,
+            ExecutionRequest::Stop => true,
+            ExecutionRequest::RunActiveStep(..) => false,
+            ExecutionRequest::ResetToFinalState => false,
+        });
+        // request reset to final state
+        (*queue_guard).push_back(ExecutionRequest::ResetToFinalState);
+        // notify
+        condvar.notify_one();
     }
 
-    fn apply_active_slot(
-        &mut self,
-        slot: Slot,
-        opt_block: Option<(BlockId, &Block)>,
-    ) -> Result<(), ExecutionError> {
-        // TODO
-        Ok(())
+    /// runs an SCE-final step (slot)
+    ///
+    /// # Arguments
+    /// * slot: target slot
+    /// * block: None if miss, Some(block_id, block) otherwise
+    fn vm_run_final_step(&mut self, slot: Slot, block: Option<(BlockId, Block)>) {
+        let (queue_lock, condvar) = &*self.execution_queue;
+        let mut queue_guard = queue_lock.lock().unwrap();
+        (*queue_guard).push_back(ExecutionRequest::RunFinalStep(ExecutionStep {
+            slot,
+            block,
+        }));
+        // notify
+        condvar.notify_one();
+    }
+
+    /// runs an SCE-active step (slot)
+    ///
+    /// # Arguments
+    /// * slot: target slot
+    /// * block: None if miss, Some(block_id, block) otherwise
+    fn vm_run_active_step(&mut self, slot: Slot, block: Option<(BlockId, Block)>) {
```

Comment:
`run_step` passing `ExecutionStep` as argument, since the caller knows what kind of step it will be.

Reply:
see prev comment, but can be discussed